src/Raven.Client/Documents/Session/InMemoryDocumentSessionOperations.cs

```diff
@@ -707,15 +709,45 @@ public SaveChangesData PrepareForSaveChanges()
             var result = new SaveChangesData
             {
                 Entities = new List<object>(),
-                Commands = new List<ICommandData>(_deferedCommands),
+                Commands = new List<ICommandData>(),
                 DeferredCommandsCount = _deferedCommands.Count,
                 Options = _saveChangesOptions
             };
-            _deferedCommands.Clear();
 
             PrepareForEntitiesDeletion(result, null);
             PrepareForEntitiesPuts(result);
 
+            if (result.DeferredCommandsCount == 0)
+            {
+                result.Commands.AddRange(_deferedCommands);
+            }
+            else
+            {
+                foreach (var deferedCommand in _deferedCommands)
+                {
+                    foreach (var command in result.Commands)
+                    {
+                        if (deferedCommand.Type == CommandType.AttachmentPUT)
+                        {
+                            if (command is PutAttachmentCommandData putAttachmentCommand &&
+                                ((PutAttachmentCommandData)command).Name == putAttachmentCommand.Name &&
+                                deferedCommand.Key == command.Key)
+                            {
+                                // TODO: Throw here and test.
```

Comment:
Need to complete this. Might actually be best to combine all those checks into a single function that will be call from the `StoreAttachment` and friends.